File: Week_02/java/src/graph/_841_KeysAndRooms.java
package graph;

import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * 第一遍：2020/08/31周一 ✅
 * 第二遍：2020/05/31周日
 * 第三遍：2020/06/06周六
 * 第四遍：2020/06/13周六
 */
class _841_KeysAndRooms {
    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        if (rooms == null || rooms.isEmpty()) return false;
        final int n = rooms.size();
        LinkedList<Integer> stack = new LinkedList<>();
        stack.push(0);
        boolean[] visited = new boolean[n];
        visited[0] = true;
        int count = 1;
        while (!stack.isEmpty()) {
            for (int v : rooms.get(stack.pop())) {
                if (!visited[v]) {
                    visited[v] = true;
                    stack.push(v);
                    if (++count == n) return true;
                }
            }
        }
        return count == n;
    }

    class SolutionBfs {
        public boolean canVisitAllRooms(List<List<Integer>> rooms) {
            if (rooms == null || rooms.isEmpty()) return false;
            final int n = rooms.size();
            Queue<Integer> queue = new LinkedList<>();
            queue.offer(0);
            boolean[] visited = new boolean[n];
            visited[0] = true;
            int count = 1;
            while (!queue.isEmpty()) {
                for (int v : rooms.get(queue.poll())) {
                    if (!visited[v]) {
                        visited[v] = true;
                        queue.offer(v);
                        if (++count == n) return true;
                    }
                }
            }
            return count == n;
        }
    }
}

File: Week_02/kotlin/src/graph/_841_KeysAndRooms.kt
package graph

import java.util.*

class _841_Dfs {
    fun canVisitAllRooms(rooms: List<List<Int>>): Boolean {
        val n = rooms.size
        val visited = BooleanArray(n)
        var count = 0
        fun dfs(u: Int) {
            if (visited[u]) return
            visited[u] = true;
            count++
            for (v in rooms[u]) {
                dfs(v);
            }
        }
        dfs(0)
        return count == n
    }
}
class _841_Bfs {
    fun canVisitAllRooms(rooms: List<List<Int>>): Boolean {
        val n = rooms.size
        val visited = BooleanArray(n).apply { this[0] = true }
        var count = 1
        val queue = LinkedList<Int>().apply { offer(0) }
        while (queue.isNotEmpty()) {
            val u = queue.poll()
            for (v in rooms[u]) {
                if (!visited[v]) {
                    visited[v] = true
                    count++
                    queue.offer(v)
                }
            }
        }
        return count == n
    }
}